design-documents/service-isolation/detailed-architecture.md

```diff
@@ -0,0 +1,87 @@
+
+Auth JWT, Customer Context JWT and Store Context are all optional, however, when present in the original client request, they MUST be retransmitted along with all subsequent calls to Magento services. They should not be retransmitted with the calls to the third-party systems.
+ 
+The following sequence diagram demonstrates the process of client application obtaining authentication and customer context JWTs.
+
+![Customer authentication](img/BFF%20-%20Customer%20Authentication%20Sequence.png)
+ 
+## Query enrichment
+
+When the user visits Magento storefront for the first time, pages that support scoping (like category page) must be rendered based on default scopes.
+Client application may send query with no currency, store code and customer context. At the same time, catalog storefront API requires store code, currency and customer group to be passed as explicit arguments. 
+The solution is to have GraphQL query enriched by the GraphQL server. Enrichment logic cannot be moved to storefront gateway because of couple reasons:
+ - it requires GraphQL query to be parsed
+ - it relies on customer context arguments to be decoded. This can be done only by GraphQL server which owns corresponding encoding secret
+ 
+Data lookup for storefront API query enrichment will be done using the following sources in the specified order:
+ 1. Customer ID will always be taken from the Authentication JWT
+ 1. Customer group, customer segment and any extensions to the customer context will always be taken from the Customer Context JWT
+ 1. Data from GraphQL query body
+ 1. Data from store context headers
+ 1. Defaults from GraphQL Server cache
+ 1. Preconfigured defaults via calls to storefront services. For example, default customer group based on current store code. The list of defaults with corresponding sources MUST be configurable
+ 
+The following diagram explains a sequence of actions needed to render a product page with PWA. After the initial call caches on server, gateway and client side will make this sequence shorter.
+
+![Product details page](img/PWA%20BFF%20-%20Product%20Page%20Sequence.png)
+
+## Scalability
+
+All components must support scalability.
+
+## Extensibility
+
+Extensibility is extremely important for Magento and multi-service Magento is not an exception. 
+
+Most of the components need to be extensible, including:
+ - GraphQL Server and any custom BFFs
+ - PWA
+ - Domain services
+ - List of supported authentication mechanisms
+ - Access control list (ACL) 
+ 
+Storefront Gateway may not need to be extensible.
+
+It is important to note that services have to be redeployed after installation of extensions. It should be possible to manage ACL via admin API gateway to avoid the need of redeployment.
+
+## Open questions
+
+ 1. Should PWA be deployed in multi-service mode? Probably not because it will hurt performance and increase complexity. 
```

Comment:
We can discuss this in more details, anytime we introduce a layer in Microservices Architecture, which acts as monolithic can potentially becomes the bottle-neck for development/release cycle; we need analyze how the particular product release will impact PWA and of-course performance aspects of the application.

Reply:
Splitting PWA into multiple services would allow to have ownership boundaries for the teams and should make upgrade easier. These are the main benefits of microservice architecture.

Reply:
@jedmao what is your opinion on this?

Reply:
We would absolutely lose performance if we split it up into separate deploys, because the dependency graph would be fragmented such that webpack cannot be smart about how bundles are created.

Reply:
@jedmao @melnikovi both are valid points on their, and frankly we should not make it too complicated for the first release, but our target architecture should aim for Micro Frontends; and we need to think end to end for Microservices without introducing a monolithic frontend layer somehow.

this is interesting article 
https://medium.com/@tomsoderlund/micro-frontends-a-microservice-approach-to-front-end-web-development-f325ebdadc16

https://www.manning.com/books/micro-frontends-in-action

Reply:
Will keep it as monolith at this iteration.